core/src/main/java/org/apache/iceberg/actions/SizeBasedDataRewriter.java

```diff
@@ -84,13 +86,30 @@ private boolean shouldRewrite(List<FileScanTask> group) {
     return enoughInputFiles(group)
         || enoughContent(group)
         || tooMuchContent(group)
-        || anyTaskHasTooManyDeletes(group);
+        || anyTaskHasTooManyDeletes(group)
+        || anyTaskHasTooHighDeleteRatio(group);
   }
 
   private boolean anyTaskHasTooManyDeletes(List<FileScanTask> group) {
     return group.stream().anyMatch(this::tooManyDeletes);
   }
 
+  private boolean anyTaskHasTooHighDeleteRatio(List<FileScanTask> group) {
+    return group.stream().anyMatch(this::tooHighDeleteRatio);
+  }
+
+  private boolean tooHighDeleteRatio(FileScanTask task) {
```

Comment:
It seems like the right place to add this logic, but what if we make it more generic? Should we also apply this check to V2 position delete files that are scoped to a particular data file (`ContentFileUtil$isFileScoped`)? We can probably add up the record counts from all file scoped deletes and use that as the number of deleted records.

Reply:
Actually, do we also need to modify `filterFiles`?